File: src/processing/phase.py
"""Phase space trajectory tracking for EarthianBioSense.

Treats each moment as a point on a trajectory in a 3D manifold,
not a dot on a line. Movement matters.

Manifold coordinates: (entrainment, breath_rate_norm, amplitude_norm)

Note on terminology:
- ENTRAINMENT = breath-heart phase coupling (local sync, the grip)
- COHERENCE = trajectory integrity over time (global, computed from trajectory autocorrelation)

The manifold position uses entrainment as one axis. Coherence is a property
of the trajectory *through* the manifold, not a coordinate within it.
"""

from dataclasses import dataclass, field
from collections import deque
from typing import Optional
import math
import time

from .hrv import HRVMetrics
from .movement import (
    SoftModeInference,
    ModeHistory,
    compute_soft_mode_membership,
    detect_mode_with_hysteresis,
    generate_movement_annotation,
    compose_movement_aware_label,
    detect_rupture_oscillation,
    VELOCITY_THRESHOLD,
    ACCELERATION_THRESHOLD,
    SETTLED_THRESHOLD,
)


@dataclass
class PhaseState:
    """A single point in phase space with timestamp."""
    timestamp: float
    position: tuple[float, float, float]  # (entrainment, breath, amplitude)


@dataclass
class PhaseDynamics:
    """Full dynamics at a moment: position + movement + history."""
    timestamp: float

    # Position in 3D manifold
    position: tuple[float, float, float]  # (entrainment, breath, amplitude)

    # First derivative - direction of movement
    velocity: tuple[float, float, float]  # (dent/dt, dbreath/dt, damp/dt)
    velocity_magnitude: float

    # Second derivative magnitude - quality of movement
    curvature: float

    # Derived qualities
    stability: float              # low curvature & low velocity → high stability
    transition_proximity: float   # placeholder - defer until basins emerge from data
    history_signature: float      # path integral over window

    # Labels (backward compat)
    phase_label: str
    mode_score: float  # collapsed scalar for terminal UI

    # === Movement-preserving classification (v1.1.0) ===
    # Soft mode inference - weighted membership across modes
    soft_mode: Optional[SoftModeInference] = None

    # Movement annotation - HOW we arrived (e.g., "settling from heightened alertness")
    movement_annotation: str = ""

    # Composed label with movement context
    movement_aware_label: str = ""

    # Hysteresis state: 'unknown', 'provisional', 'established'
    mode_status: str = "unknown"

    # Dwell time in current mode (seconds)
    dwell_time: float = 0.0

    # Second derivative of mode_score (scalar), NOT 3D trajectory acceleration.
    # The 3D trajectory acceleration is stored in the `curvature` field.
    mode_score_acceleration: float = 0.0


# Default/empty dynamics for cold start
EMPTY_DYNAMICS = PhaseDynamics(
    timestamp=0,
    position=(0.0, 0.0, 0.0),
    velocity=(0.0, 0.0, 0.0),
    velocity_magnitude=0.0,
    curvature=0.0,
    stability=0.0,
    transition_proximity=0.0,
    history_signature=0.0,
    phase_label="initializing",
    mode_score=0.0
)


class PhaseTrajectory:
    """Rolling buffer of PhaseStates with dynamics computation.

    Tracks trajectory through 3D manifold and computes:
    - Velocity (finite difference)
    - Curvature (second derivative magnitude)
    - Stability (inverse of movement intensity)
    - History signature (path integral)
    - Movement-preserving mode classification (v1.1.0)
    """

    def __init__(self, window_size: int = 30):
        """Initialize trajectory buffer.

        Args:
            window_size: Number of states to keep. At ~1 state/second,
                        30 = ~30 seconds of trajectory history.
        """
        self.states: deque[PhaseState] = deque(maxlen=window_size)
        self.cumulative_path_length: float = 0.0
        self._last_velocity: tuple[float, float, float] = (0.0, 0.0, 0.0)

        # Movement-preserving classification state (v1.1.0)
        self.mode_history: ModeHistory = ModeHistory()
        self._last_soft_inference: Optional[SoftModeInference] = None
        self._last_mode_score: float = 0.0
        self._mode_score_velocity: float = 0.0

    def append(self, metrics: HRVMetrics, timestamp: float | None = None) -> PhaseDynamics:
        """Add new state from HRV metrics, compute dynamics.

        Args:
            metrics: Current HRV metrics snapshot
            timestamp: Optional timestamp (uses time.time() if not provided)

        Returns:
            PhaseDynamics with full trajectory analysis
        """
        if timestamp is None:
            timestamp = time.time()

        position = self._metrics_to_position(metrics)
        new_state = PhaseState(timestamp, position)

        # Compute dynamics before adding (need previous states)
        dynamics = self._compute_dynamics(new_state, metrics)

        # Update path integral
        if len(self.states) > 0:
            prev = self.states[-1]
            step_distance = self._euclidean_distance(prev.position, position)
            self.cumulative_path_length += step_distance

        self.states.append(new_state)

        return dynamics

    def _metrics_to_position(self, m: HRVMetrics) -> tuple[float, float, float]:
        """Map HRV metrics to manifold coordinates (all normalized 0-1)."""
        # Entrainment: already 0-1
        ent = m.entrainment

        # Breath rate: normalize ~4-20 breaths/min → 0-1
        # 4 bpm → 0, 12 bpm → 0.5, 20 bpm → 1
        if m.breath_rate is not None:
            breath = max(0.0, min(1.0, (m.breath_rate - 4) / 16))
        else:
            breath = 0.5  # default to middle if unknown

        # Amplitude: normalize 0-200ms → 0-1
        amp = min(1.0, m.amplitude / 200)

        return (ent, breath, amp)

    def _compute_dynamics(self, new_state: PhaseState, metrics: HRVMetrics) -> PhaseDynamics:
        """Compute velocity, curvature, stability from trajectory."""

        if len(self.states) < 2:
            # Not enough history - return minimal dynamics
            # Still compute soft mode for early feedback
            soft_mode = compute_soft_mode_membership(
                entrainment=metrics.entrainment,
                breath_steady=metrics.breath_steady,
                amp_norm=min(1.0, metrics.amplitude / 200),
                volatility=metrics.rr_volatility,
                previous_inference=self._last_soft_inference
            )
            self._last_soft_inference = soft_mode
            self._last_mode_score = metrics.mode_score

            return PhaseDynamics(
                timestamp=new_state.timestamp,
                position=new_state.position,
                velocity=(0.0, 0.0, 0.0),
                velocity_magnitude=0.0,
                curvature=0.0,
                stability=0.5,  # neutral
                transition_proximity=0.0,
                history_signature=0.0,
                phase_label="warming up",
                mode_score=metrics.mode_score,
                soft_mode=soft_mode,
                movement_annotation="insufficient data",
                movement_aware_label=soft_mode.primary_mode,
                mode_status="unknown",
                dwell_time=0.0,
                mode_score_acceleration=0.0
            )

        # Get recent states for derivative computation
        prev = self.states[-1]
        prev_prev = self.states[-2] if len(self.states) >= 2 else prev

        # Time deltas
        dt1 = new_state.timestamp - prev.timestamp
        dt2 = prev.timestamp - prev_prev.timestamp

        # Avoid division by zero
        dt1 = max(dt1, 0.001)
        dt2 = max(dt2, 0.001)

        # Velocity: finite difference (first derivative)
        velocity = tuple(
            (new_state.position[i] - prev.position[i]) / dt1
            for i in range(3)
        )
        velocity_magnitude = self._vector_magnitude(velocity)

        # Previous velocity for curvature
        prev_velocity = tuple(
            (prev.position[i] - prev_prev.position[i]) / dt2
            for i in range(3)
        )

        # Curvature: magnitude of acceleration (second derivative)
        # Using average dt for acceleration computation
        dt_avg = (dt1 + dt2) / 2
        acceleration = tuple(
            (velocity[i] - prev_velocity[i]) / dt_avg
            for i in range(3)
        )
        curvature = self._vector_magnitude(acceleration)

        # Store for next iteration
        self._last_velocity = velocity

        # Stability: inverse relationship with movement intensity
        # High velocity + high curvature = low stability
        movement_intensity = velocity_magnitude + curvature * 0.5
        stability = 1.0 / (1.0 + movement_intensity * 2)
        stability = max(0.0, min(1.0, stability))

        # History signature: windowed path integral (distance within rolling buffer)
        # Uses only the rolling window, not cumulative session path, to avoid
        # saturating to 1.0 within minutes (P1-B fix from RAA-EBS-001).
        if len(self.states) >= 1:
            windowed_path = sum(
                self._euclidean_distance(
                    self.states[i].position, self.states[i + 1].position
                )
                for i in range(len(self.states) - 1)
            )
            windowed_path += self._euclidean_distance(
                self.states[-1].position, new_state.position
            )
            window_time = new_state.timestamp - self.states[0].timestamp
            window_time = max(window_time, 1.0)
            history_signature = min(1.0, windowed_path / window_time / 0.5)
        else:
            history_signature = 0.0

        # Transition proximity: defer for now
        # Will need basin definitions to compute meaningfully
        transition_proximity = 0.0

        # Phase label from dynamics (not just thresholds)
        phase_label = self._infer_phase_label(
            new_state.position, velocity_magnitude, curvature, stability
        )

        # === Movement-preserving classification (v1.1.0) ===

        # Compute soft mode membership
        amp_norm = min(1.0, metrics.amplitude / 200)
        soft_mode = compute_soft_mode_membership(
            entrainment=metrics.entrainment,
            breath_steady=metrics.breath_steady,
            amp_norm=amp_norm,
            volatility=metrics.rr_volatility,
            previous_inference=self._last_soft_inference
        )

        # Detect mode with hysteresis
        detected_mode, mode_confidence, mode_meta = detect_mode_with_hysteresis(
            soft_inference=soft_mode,
            mode_history=self.mode_history,
            timestamp=new_state.timestamp
        )

        # Compute mode_score velocity and acceleration for movement annotation
        mode_score_velocity = (metrics.mode_score - self._last_mode_score) / dt1
        mode_score_accel = (mode_score_velocity - self._mode_score_velocity) / dt1

        # Generate movement annotation
        movement_annotation = generate_movement_annotation(
            velocity_magnitude=abs(mode_score_velocity),
            mode_score_acceleration=mode_score_accel,
            previous_mode=mode_meta.get('previous_mode'),
            dwell_time=mode_meta.get('dwell_time', 0.0)
        )

        # Compose movement-aware label
        movement_aware_label = compose_movement_aware_label(detected_mode, movement_annotation)

        # Update history for next iteration
        self.mode_history.append(detected_mode, mode_confidence, new_state.timestamp)
        self._last_soft_inference = soft_mode
        self._last_mode_score = metrics.mode_score
        self._mode_score_velocity = mode_score_velocity

        return PhaseDynamics(
            timestamp=new_state.timestamp,
            position=new_state.position,
            velocity=velocity,
            velocity_magnitude=velocity_magnitude,
            curvature=curvature,
            stability=stability,
            transition_proximity=transition_proximity,
            history_signature=history_signature,
            phase_label=phase_label,
            mode_score=metrics.mode_score,  # keep original for backward compat
            soft_mode=soft_mode,
            movement_annotation=movement_annotation,
            movement_aware_label=movement_aware_label,
            mode_status=mode_meta.get('state_status', 'unknown'),
            dwell_time=mode_meta.get('dwell_time', 0.0),
            mode_score_acceleration=abs(mode_score_accel)
        )

    def _infer_phase_label(
        self,
        position: tuple[float, float, float],
        velocity_mag: float,
        curvature: float,
        stability: float
    ) -> str:
        """Infer phase label from dynamics, not just position thresholds.

        This is still provisional - letting patterns emerge rather than
        imposing ontology.

        Note: 'ent' here is entrainment (breath-heart sync), not coherence.
        Coherence (trajectory integrity) would require looking at the
        trajectory history, not just current position.
        """
        ent, breath, amp = position

        # High stability + high entrainment = settled/entrained dwelling
        if stability > 0.7 and ent > 0.6:
            return "entrained dwelling"

        # High curvature = turning point, transition
        if curvature > 0.3:
            if ent > 0.5:
                return "inflection (from entrainment)"
            else:
                return "inflection (seeking)"

        # High velocity + direction
        if velocity_mag > 0.1:
            if ent > 0.5:
                return "flowing (entrained)"
            else:
                return "active transition"

        # Low everything = dwelling but where?
        if stability > 0.6:
            if ent > 0.5:
                return "settling into entrainment"
            elif ent > 0.3:
                return "neutral dwelling"
            else:
                return "alert stillness"  # v1.1.0: "vigilant" -> "alert"

        # Default: transitional
        return "transitional"

    @staticmethod
    def _euclidean_distance(p1: tuple[float, ...], p2: tuple[float, ...]) -> float:
        """Compute Euclidean distance between two points."""
        return math.sqrt(sum((a - b) ** 2 for a, b in zip(p1, p2)))

    @staticmethod
    def _vector_magnitude(v: tuple[float, ...]) -> float:
        """Compute magnitude of a vector."""
        return math.sqrt(sum(x ** 2 for x in v))

    def get_recent_trajectory(self, n: int = 10) -> list[PhaseState]:
        """Get the n most recent states for visualization/analysis."""
        return list(self.states)[-n:]

    def reset(self):
        """Clear trajectory buffer and path integral."""
        self.states.clear()
        self.cumulative_path_length = 0.0
        self._last_velocity = (0.0, 0.0, 0.0)
        # Reset movement-preserving state (v1.1.0)
        self.mode_history.clear()
        self._last_soft_inference = None
        self._last_mode_score = 0.0
        self._mode_score_velocity = 0.0

    def compute_trajectory_coherence(self, lag: int = 5) -> float:
        """
        Compute COHERENCE as trajectory autocorrelation.

        This is the key insight: coherence is NOT entrainment (breath-heart sync).
        Coherence is how well the trajectory through phase space hangs together
        over time — the autocorrelation of movement patterns.

        High coherence = the system's movement has integrity, patterns persist
        Low coherence = fragmented, reactive, no trajectory continuity

        Args:
            lag: Number of states to lag for autocorrelation (default 5 = ~5 seconds)

        Returns:
            coherence: 0-1 score of trajectory integrity
        """
        if len(self.states) < lag + 3:
            return 0.0  # Insufficient data

        # Extract velocity vectors from recent trajectory
        # We're asking: does the *pattern of movement* correlate with itself?
        positions = [s.position for s in self.states]

        # Compute velocity sequence (first differences)
        velocities = []
        for i in range(1, len(positions)):
            v = tuple(positions[i][j] - positions[i-1][j] for j in range(3))
            velocities.append(v)

        if len(velocities) < lag + 2:
            return 0.0

        # Compute autocorrelation of velocity magnitudes
        # This captures: is the *intensity* of movement consistent over time?
        v_mags = [self._vector_magnitude(v) for v in velocities]

        n = len(v_mags)
        mean_v = sum(v_mags) / n
        variance = sum((x - mean_v) ** 2 for x in v_mags) / n

        if variance < 1e-10:
            # Near-zero variance = perfectly still = high coherence (dwelling)
            return 0.8

        # Autocovariance at lag
        # Use n as denominator (matching variance above) for consistent normalization.
        # Using (n - lag) inflates the result by n/(n-lag) at small buffer sizes.
        autocovariance = sum(
            (v_mags[i] - mean_v) * (v_mags[i + lag] - mean_v)
            for i in range(n - lag)
        ) / n

        autocorr = autocovariance / variance

        # Also consider direction consistency (are we moving in consistent directions?)
        # Compute cosine similarity between velocity vectors at lag
        direction_coherence = 0.0
        count = 0
        for i in range(len(velocities) - lag):
            v1 = velocities[i]
            v2 = velocities[i + lag]
            mag1 = self._vector_magnitude(v1)
            mag2 = self._vector_magnitude(v2)
            if mag1 > 1e-6 and mag2 > 1e-6:
                dot = sum(v1[j] * v2[j] for j in range(3))
                cosine = dot / (mag1 * mag2)
                direction_coherence += (cosine + 1) / 2  # Normalize to 0-1
                count += 1

        if count > 0:
            direction_coherence /= count
        else:
            direction_coherence = 0.5  # Neutral if no movement

        # Combine magnitude autocorrelation and direction consistency
        # Both matter: coherent trajectory has consistent intensity AND direction patterns
        coherence = 0.5 * max(0.0, autocorr) + 0.5 * direction_coherence

        return max(0.0, min(1.0, coherence))

File: desktop/src-tauri/src/hrv/phase.rs
//! Phase-space trajectory tracking for HRV analysis.
//!
//! Ported from `src/processing/phase.py`. Maps HRV metrics into a 3D phase
//! space (entrainment, breath_norm, amplitude_norm) and tracks dynamics —
//! velocity, curvature, stability, history signature — over a rolling window.

use std::collections::VecDeque;

use serde::Serialize;

use super::HRVMetrics;
use super::movement::{
    ModeHistory, SoftModeInference, compose_movement_aware_label,
    compute_soft_mode_membership, detect_mode_with_hysteresis,
    generate_movement_annotation,
};

// ---------------------------------------------------------------------------
// Data structures
// ---------------------------------------------------------------------------

/// A single snapshot in the 3D phase space.
#[derive(Debug, Clone)]
pub struct PhaseState {
    pub timestamp: f64,
    /// [entrainment, breath_norm, amplitude_norm]
    pub position: [f64; 3],
}

/// Full dynamics output for one time-step.
#[derive(Debug, Clone, Serialize)]
pub struct PhaseDynamics {
    pub timestamp: f64,
    pub position: [f64; 3],
    pub velocity: [f64; 3],
    pub velocity_magnitude: f64,
    pub curvature: f64,
    pub stability: f64,
    pub transition_proximity: f64,
    pub history_signature: f64,
    pub phase_label: String,
    pub mode_score: f64,
    pub soft_mode: Option<SoftModeInference>,
    pub movement_annotation: String,
    pub movement_aware_label: String,
    pub mode_status: String,
    pub dwell_time: f64,
    pub mode_score_acceleration: f64,
}

// ---------------------------------------------------------------------------
// PhaseTrajectory — rolling buffer with dynamics computation
// ---------------------------------------------------------------------------

pub struct PhaseTrajectory {
    states: VecDeque<PhaseState>,
    window_size: usize,
    cumulative_path_length: f64,
    last_velocity: [f64; 3],
    mode_history: ModeHistory,
    last_soft_inference: Option<SoftModeInference>,
    last_mode_score: f64,
    mode_score_velocity: f64,
}

impl PhaseTrajectory {
    pub fn new(window_size: usize) -> Self {
        Self {
            states: VecDeque::with_capacity(window_size),
            window_size,
            cumulative_path_length: 0.0,
            last_velocity: [0.0; 3],
            mode_history: ModeHistory::new(100),
            last_soft_inference: None,
            last_mode_score: 0.0,
            mode_score_velocity: 0.0,
        }
    }

    pub fn append(&mut self, metrics: &HRVMetrics, timestamp: f64) -> PhaseDynamics {
        let position = Self::metrics_to_position(metrics);
        let new_state = PhaseState { timestamp, position };
        let dynamics = self.compute_dynamics(&new_state, metrics);

        if let Some(prev) = self.states.back() {
            self.cumulative_path_length +=
                euclidean_distance(&prev.position, &new_state.position);
        }

        if self.states.len() >= self.window_size {
            self.states.pop_front();
        }
        self.states.push_back(new_state);

        dynamics
    }

    pub fn metrics_to_position(m: &HRVMetrics) -> [f64; 3] {
        let entrainment = m.entrainment;
        let breath_norm = match m.breath_rate {
            Some(rate) => ((rate - 4.0) / 16.0).clamp(0.0, 1.0),
            None => 0.5,
        };
        let amplitude_norm = (m.amplitude as f64 / 200.0).min(1.0);
        [entrainment, breath_norm, amplitude_norm]
    }

    fn compute_dynamics(
        &mut self,
        new_state: &PhaseState,
        metrics: &HRVMetrics,
    ) -> PhaseDynamics {
        let amp_norm = (metrics.amplitude as f64 / 200.0).min(1.0);

        // Soft mode — always computed
        let soft_mode = compute_soft_mode_membership(
            metrics.entrainment,
            metrics.breath_steady,
            amp_norm,
            metrics.rr_volatility,
            0.2, // temperature
            self.last_soft_inference.as_ref(),
        );

        if self.states.len() < 2 {
            // Warming up — not enough history for derivatives
            let (detected_mode, _confidence, meta) = detect_mode_with_hysteresis(
                &soft_mode,
                &mut self.mode_history,
                new_state.timestamp,
            );

            let previous_mode = meta.get("previous_mode").map(|s| s.as_str());
            let dwell_time: f64 = meta
                .get("dwell_time")
                .and_then(|s| s.parse().ok())
                .unwrap_or(0.0);
            let mode_status = meta
                .get("state_status")
                .cloned()
                .unwrap_or_else(|| "unknown".to_string());

            let movement_annotation = generate_movement_annotation(
                None, // no velocity yet
                None,
                previous_mode,
                dwell_time,
            );

            let movement_aware_label =
                compose_movement_aware_label(&detected_mode, &movement_annotation);

            self.mode_history
                .append(&detected_mode, soft_mode.membership[&soft_mode.primary_mode], new_state.timestamp);
            self.last_soft_inference = Some(soft_mode.clone());
            self.last_mode_score = metrics.mode_score;

            return PhaseDynamics {
                timestamp: new_state.timestamp,
                position: new_state.position,
                velocity: [0.0; 3],
                velocity_magnitude: 0.0,
                curvature: 0.0,
                stability: 0.5,
                transition_proximity: 0.0,
                history_signature: 0.0,
                phase_label: "warming up".to_string(),
                mode_score: metrics.mode_score,
                soft_mode: Some(soft_mode),
                movement_annotation,
                movement_aware_label,
                mode_status,
                dwell_time: 0.0,
                mode_score_acceleration: 0.0,
            };
        }

        // --- Velocity (finite difference) ---
        let prev = &self.states[self.states.len() - 1];
        let dt = (new_state.timestamp - prev.timestamp).max(0.001);

        let velocity = [
            (new_state.position[0] - prev.position[0]) / dt,
            (new_state.position[1] - prev.position[1]) / dt,
            (new_state.position[2] - prev.position[2]) / dt,
        ];
        let velocity_mag = vector_magnitude(&velocity);

        // --- Curvature (magnitude of acceleration) ---
        let acceleration = [
            (velocity[0] - self.last_velocity[0]) / dt,
            (velocity[1] - self.last_velocity[1]) / dt,
            (velocity[2] - self.last_velocity[2]) / dt,
        ];
        let curvature = vector_magnitude(&acceleration);

        // --- Stability ---
        let stability =
            (1.0 / (1.0 + (velocity_mag + curvature * 0.5) * 2.0)).clamp(0.0, 1.0);

        // --- History signature (rolling window only — RAA-EBS-001 fix) ---
        let history_signature = self.compute_rolling_path_signature(new_state);

        // --- Phase label ---
        let phase_label = infer_phase_label(
            &new_state.position,
            velocity_mag,
            curvature,
            stability,
        );

        let transition_proximity = 1.0 - stability;

        // --- Mode dynamics ---
        let (detected_mode, _confidence, meta) = detect_mode_with_hysteresis(
            &soft_mode,
            &mut self.mode_history,
            new_state.timestamp,
        );

        let previous_mode = meta.get("previous_mode").map(|s| s.as_str());
        let dwell_time: f64 = meta
            .get("dwell_time")
            .and_then(|s| s.parse().ok())
            .unwrap_or(0.0);
        let mode_status = meta
            .get("state_status")
            .cloned()
            .unwrap_or_else(|| "unknown".to_string());

        // --- Mode score acceleration ---
        let prev_mode_score_velocity = self.mode_score_velocity;
        self.mode_score_velocity = (metrics.mode_score - self.last_mode_score) / dt;
        let mode_score_accel =
            ((self.mode_score_velocity - prev_mode_score_velocity) / dt).abs();

        // --- Movement annotation ---
        let movement_annotation = generate_movement_annotation(
            Some(self.mode_score_velocity.abs()),
            Some(mode_score_accel),
            previous_mode,
            dwell_time,
        );

        let movement_aware_label =
            compose_movement_aware_label(&detected_mode, &movement_annotation);

        // --- Update internal state ---
        self.mode_history
            .append(&detected_mode, _confidence, new_state.timestamp);
        self.last_velocity = velocity;
        self.last_soft_inference = Some(soft_mode.clone());
        self.last_mode_score = metrics.mode_score;

        PhaseDynamics {
            timestamp: new_state.timestamp,
            position: new_state.position,
            velocity,
            velocity_magnitude: velocity_mag,
            curvature,
            stability,
            transition_proximity,
            history_signature,
            phase_label,
            mode_score: metrics.mode_score,
            soft_mode: Some(soft_mode),
            movement_annotation,
            movement_aware_label,
            mode_status,
            dwell_time,
            mode_score_acceleration: mode_score_accel,
        }
    }

    fn compute_rolling_path_signature(&self, new_state: &PhaseState) -> f64 {
        if self.states.len() < 2 {
            return 0.0;
        }
        let mut path_length = 0.0;
        for i in 1..self.states.len() {
            path_length += euclidean_distance(
                &self.states[i - 1].position,
                &self.states[i].position,
            );
        }
        if let Some(last) = self.states.back() {
            path_length += euclidean_distance(&last.position, &new_state.position);
        }
        let window_time =
            (new_state.timestamp - self.states[0].timestamp).max(0.001);
        ((path_length / window_time) / 0.5).min(1.0)
    }

    pub fn compute_trajectory_coherence(&self, lag: usize) -> f64 {
        if self.states.len() < lag + 2 {
            return 0.0;
        }

        let n = self.states.len() - 1;
        let mut vel_mags: Vec<f64> = Vec::with_capacity(n);
        let mut vel_dirs: Vec<[f64; 3]> = Vec::with_capacity(n);

        for i in 0..n {
            let dt = (self.states[i + 1].timestamp - self.states[i].timestamp)
                .max(0.001);
            let v = [
                (self.states[i + 1].position[0] - self.states[i].position[0]) / dt,
                (self.states[i + 1].position[1] - self.states[i].position[1]) / dt,
                (self.states[i + 1].position[2] - self.states[i].position[2]) / dt,
            ];
            vel_mags.push(vector_magnitude(&v));
            vel_dirs.push(v);
        }

        let mean_mag: f64 = vel_mags.iter().sum::<f64>() / vel_mags.len() as f64;
        let variance: f64 = vel_mags
            .iter()
            .map(|m| (m - mean_mag).powi(2))
            .sum::<f64>()
            / vel_mags.len() as f64;

        if variance < 1e-10 {
            return 0.8;
        }

        // Autocorrelation with n as denominator
        let autocorr = {
            let mut num = 0.0;
            for i in 0..(vel_mags.len() - lag) {
                num += (vel_mags[i] - mean_mag) * (vel_mags[i + lag] - mean_mag);
            }
            num / (vel_mags.len() as f64 * variance)
        };

        // Direction consistency via cosine similarity
        let direction_coherence = {
            let mut cos_sum = 0.0;
            let mut count = 0usize;
            for i in 0..(vel_dirs.len() - lag) {
                let a = &vel_dirs[i];
                let b = &vel_dirs[i + lag];
                let mag_a = vector_magnitude(a);
                let mag_b = vector_magnitude(b);
                if mag_a > 1e-10 && mag_b > 1e-10 {
                    let dot = a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
                    cos_sum += (dot / (mag_a * mag_b) + 1.0) / 2.0;
                    count += 1;
                }
            }
            if count == 0 { 0.5 } else { cos_sum / count as f64 }
        };

        let coherence = 0.5 * autocorr.max(0.0) + 0.5 * direction_coherence;
        coherence.clamp(0.0, 1.0)
    }

    pub fn reset(&mut self) {
        self.states.clear();
        self.cumulative_path_length = 0.0;
        self.last_velocity = [0.0; 3];
        self.mode_history = ModeHistory::new(100);
        self.last_soft_inference = None;
        self.last_mode_score = 0.0;
        self.mode_score_velocity = 0.0;
    }
}

// ---------------------------------------------------------------------------
// Phase label inference
// ---------------------------------------------------------------------------

fn infer_phase_label(
    position: &[f64; 3],
    velocity_mag: f64,
    curvature: f64,
    stability: f64,
) -> String {
    let ent = position[0];

    if stability > 0.7 && ent > 0.6 {
        "entrained dwelling".to_string()
    } else if curvature > 0.3 && ent > 0.5 {
        "inflection (from entrainment)".to_string()
    } else if curvature > 0.3 {
        "inflection (seeking)".to_string()
    } else if velocity_mag > 0.1 && ent > 0.5 {
        "flowing (entrained)".to_string()
    } else if velocity_mag > 0.1 {
        "active transition".to_string()
    } else if stability > 0.6 && ent > 0.5 {
        "settling into entrainment".to_string()
    } else if stability > 0.6 && ent > 0.3 {
        "neutral dwelling".to_string()
    } else if stability > 0.6 {
        "alert stillness".to_string()
    } else {
        "transitional".to_string()
    }
}

// ---------------------------------------------------------------------------
// Helper functions
// ---------------------------------------------------------------------------

pub fn euclidean_distance(p1: &[f64; 3], p2: &[f64; 3]) -> f64 {
    let dx = p1[0] - p2[0];
    let dy = p1[1] - p2[1];
    let dz = p1[2] - p2[2];
    (dx * dx + dy * dy + dz * dz).sqrt()
}

pub fn vector_magnitude(v: &[f64; 3]) -> f64 {
    (v[0] * v[0] + v[1] * v[1] + v[2] * v[2]).sqrt()
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::hrv::compute_hrv_metrics;

    fn make_metrics(
        entrainment: f64,
        breath_rate: Option<f64>,
        amplitude: u16,
        mode_score: f64,
    ) -> HRVMetrics {
        HRVMetrics {
            mean_rr: 800.0,
            min_rr: 750,
            max_rr: 750 + amplitude,
            amplitude,
            entrainment,
            entrainment_label: "[low]".to_string(),
            breath_rate,
            breath_steady: true,
            rr_volatility: 0.03,
            mode_label: "transitional".to_string(),
            mode_score,
        }
    }

    #[test]
    fn test_metrics_to_position() {
        let m = make_metrics(0.8, Some(12.0), 100, 0.5);
        let pos = PhaseTrajectory::metrics_to_position(&m);
        assert!((pos[0] - 0.8).abs() < 1e-9);
        assert!((pos[1] - 0.5).abs() < 1e-9); // (12-4)/16
        assert!((pos[2] - 0.5).abs() < 1e-9); // 100/200

        let m2 = make_metrics(0.3, None, 250, 0.0);
        let pos2 = PhaseTrajectory::metrics_to_position(&m2);
        assert!((pos2[1] - 0.5).abs() < 1e-9); // None -> 0.5
        assert!((pos2[2] - 1.0).abs() < 1e-9); // clamped
    }

    #[test]
    fn test_phase_trajectory_warmup() {
        let mut traj = PhaseTrajectory::new(30);
        let m = make_metrics(0.5, Some(10.0), 80, 0.4);
        let d1 = traj.append(&m, 1.0);
        assert_eq!(d1.phase_label, "warming up");
        let d2 = traj.append(&m, 2.0);
        assert_eq!(d2.phase_label, "warming up");
        // Third has 2 states in buffer — real label
        let d3 = traj.append(&m, 3.0);
        assert_ne!(d3.phase_label, "warming up");
    }

    #[test]
    fn test_velocity_computation() {
        let mut traj = PhaseTrajectory::new(30);
        let m1 = make_metrics(0.0, Some(4.0), 0, 0.0);
        traj.append(&m1, 0.0);
        traj.append(&m1, 1.0);
        let m2 = make_metrics(0.5, Some(4.0), 0, 0.0);
        let d = traj.append(&m2, 2.0);
        assert!(
            (d.velocity[0] - 0.5).abs() < 1e-6,
            "expected 0.5, got {}",
            d.velocity[0]
        );
    }

    #[test]
    fn test_stability_high_when_still() {
        let mut traj = PhaseTrajectory::new(30);
        let m = make_metrics(0.7, Some(10.0), 100, 0.5);
        for i in 0..10 {
            traj.append(&m, i as f64);
        }
        let d = traj.append(&m, 10.0);
        assert!(d.stability > 0.9, "got {}", d.stability);
    }

    #[test]
    fn test_coherence_insufficient_data() {
        let traj = PhaseTrajectory::new(30);
        assert!((traj.compute_trajectory_coherence(1)).abs() < 1e-9);
    }

    #[test]
    fn test_phase_labels() {
        assert_eq!(infer_phase_label(&[0.8, 0.5, 0.5], 0.0, 0.0, 0.9), "entrained dwelling");
        assert_eq!(infer_phase_label(&[0.6, 0.5, 0.5], 0.0, 0.5, 0.3), "inflection (from entrainment)");
        assert_eq!(infer_phase_label(&[0.3, 0.5, 0.5], 0.0, 0.5, 0.3), "inflection (seeking)");
        assert_eq!(infer_phase_label(&[0.6, 0.5, 0.5], 0.2, 0.1, 0.3), "flowing (entrained)");
        assert_eq!(infer_phase_label(&[0.3, 0.5, 0.5], 0.2, 0.1, 0.3), "active transition");
        assert_eq!(infer_phase_label(&[0.6, 0.5, 0.5], 0.05, 0.1, 0.65), "settling into entrainment");
        assert_eq!(infer_phase_label(&[0.4, 0.5, 0.5], 0.05, 0.1, 0.65), "neutral dwelling");
        assert_eq!(infer_phase_label(&[0.2, 0.5, 0.5], 0.05, 0.1, 0.65), "alert stillness");
        assert_eq!(infer_phase_label(&[0.4, 0.5, 0.5], 0.05, 0.1, 0.5), "transitional");
    }
}
